vci-service-impl/src/main/java/io/mosip/certify/vci/pop/JwtProofValidator.java

```diff
@@ -0,0 +1,172 @@
+            claimsSetVerifier.setMaxClockSkew(0);
+
+            JWSKeySelector keySelector = new JWSVerificationKeySelector(allowedSignatureAlgorithms,
+                    new ImmutableJWKSet(new JWKSet(jwk)));
+            ConfigurableJWTProcessor jwtProcessor = new DefaultJWTProcessor();
+            jwtProcessor.setJWSKeySelector(keySelector);
+            jwtProcessor.setJWSTypeVerifier(new DefaultJOSEObjectTypeVerifier(new JOSEObjectType(HEADER_TYP)));
+            jwtProcessor.setJWTClaimsSetVerifier(claimsSetVerifier);
+            jwtProcessor.process(credentialProof.getJwt(), null);
+            return true;
+        } catch (InvalidRequestException e) {
+            log.error("Invalid proof : {}", e.getErrorCode());
+        }  catch (ParseException e) {
+            log.error("Failed to parse jwt in the credential proof", e);
+        } catch (BadJOSEException | JOSEException e) {
+            log.error("JWT proof verification failed", e);
+        }
+        return false;
+    }
+
+    @Override
+    public String getKeyMaterial(CredentialProof credentialProof) {
+        try {
+            SignedJWT jwt = (SignedJWT) JWTParser.parse(credentialProof.getJwt());
+            JWK jwk = getKeyFromHeader(jwt.getHeader());
+            byte[] keyBytes = jwk.toJSONString().getBytes(StandardCharsets.UTF_8);
+            return DID_JWK_PREFIX.concat(Base64.getUrlEncoder().encodeToString(keyBytes));
+        } catch (ParseException e) {
+            log.error("Failed to parse jwt in the credential proof", e);
+        }
+        throw new InvalidRequestException(ErrorConstants.PROOF_HEADER_INVALID_KEY);
+    }
+
+    private void validateHeaderClaims(JWSHeader jwsHeader) {
+        if(Objects.isNull(jwsHeader.getType()) || !HEADER_TYP.equals(jwsHeader.getType().getType()))
+            throw new InvalidRequestException(ErrorConstants.PROOF_HEADER_INVALID_TYP);
+
+        if(Objects.isNull(jwsHeader.getAlgorithm()) || !supportedAlgorithms.contains(jwsHeader.getAlgorithm().getName()))
+            throw new InvalidRequestException(ErrorConstants.PROOF_HEADER_INVALID_ALG);
+
+        if(Objects.isNull(jwsHeader.getKeyID()) && Objects.isNull(jwsHeader.getJWK()))
+            throw new InvalidRequestException(ErrorConstants.PROOF_HEADER_INVALID_KEY);
+
+        //both cannot be present, either one of them is only allowed
+        if(Objects.nonNull(jwsHeader.getKeyID()) && Objects.nonNull(jwsHeader.getJWK()))
+            throw new InvalidRequestException(ErrorConstants.PROOF_HEADER_AMBIGUOUS_KEY);
+
+        //TODO x5c and trust_chain validation
+    }
+
+    private JWK getKeyFromHeader(JWSHeader jwsHeader) {
+        if(Objects.nonNull(jwsHeader.getJWK()))
+            return jwsHeader.getJWK();
+
+        return resolveDID(jwsHeader.getKeyID());
+    }
+
+    /**
+     * Currently only handles did:jwk, Need to handle other methods
+     * @param keyId
+     * @return
+     */
+    private JWK resolveDID(String did) {
```

Comment:
Can we create a task to make this into a factory implementation to support various forms of DID ? Also we should check if there is a library from DIF / W3C which can do this, instead of we custom handling this.

Reply:
created a task for [this](https://mosip.atlassian.net/browse/INJICERT-207)